corehq/apps/hqadmin/reports.py

```diff
@@ -333,3 +337,146 @@ def _shorten_and_hyperlink_commit(self, commit_sha):
+            return []
+
+        return ucr_domains
+
+    @property
+    def show_ff_enabled_domains(self):
+        return self.restriction_ff_status == UCRRestrictionFFStatus.Enabled.name
+
+    @property
+    def show_ff_disabled_domains(self):
+        return self.restriction_ff_status == UCRRestrictionFFStatus.NotEnabled.name
+
+    @property
+    def show_should_enable_ff_domains(self):
+        return self.restriction_ff_status == UCRRestrictionFFStatus.ShouldEnable.name
+
+    @property
+    def show_should_disable_ff_domains(self):
+        return self.restriction_ff_status == UCRRestrictionFFStatus.CanDisable.name
+
+    @property
+    def show_all_domains(self):
+        return not self.restriction_ff_status
+
+    def should_show_domain(self, domain, total_cases, total_forms):
+        if self.show_all_domains:
+            return True
+
+        should_restrict_rebuild = self._should_restrict_rebuild(total_cases, total_forms)
+        restriction_ff_enabled = self._rebuild_restricted_ff_enabled(domain)
+
+        if self.show_ff_enabled_domains:
+            return restriction_ff_enabled
+        if self.show_ff_disabled_domains:
+            return not restriction_ff_enabled
+        if self.show_should_enable_ff_domains:
+            return should_restrict_rebuild and not restriction_ff_enabled
+        if self.show_should_disable_ff_domains:
+            return not should_restrict_rebuild and restriction_ff_enabled
+
+    def _row_data(self, domain, case_count, form_count):
+        return [
+            domain,
+            case_count,
+            form_count,
+            self._restrict_rebuild_column_data(domain, case_count, form_count),
+        ]
+
+    def _restrict_rebuild_column_data(self, domain, case_count, form_count):
+        from django.utils.safestring import mark_safe
+        from corehq.apps.toggle_ui.views import ToggleEditView
+
+        restriction_ff_enabled = self._rebuild_restricted_ff_enabled(domain)
+        toggle_edit_url = reverse(ToggleEditView.urlname, args=(RESTRICT_DATA_SOURCE_REBUILD.slug,))
+
+        if self._should_restrict_rebuild(case_count, form_count):
+            if not restriction_ff_enabled:
+                return mark_safe(f"""
+                    <a href={toggle_edit_url}>{gettext_lazy("Rebuild restriction required")}</a>
+                """)
+            return gettext_lazy("Rebuild restricted")
+
+        if restriction_ff_enabled:
+            return mark_safe(f"""
+                <a href={toggle_edit_url}>{gettext_lazy("Rebuild restriction not required")}</a>
+            """)
+        return gettext_lazy("No rebuild restriction required")
+
+    def _should_restrict_rebuild(self, case_count, form_count):
+        return case_count >= self.UCR_RESTRICTION_THRESHOLD or form_count >= self.UCR_RESTRICTION_THRESHOLD
+
+    @staticmethod
+    def _rebuild_restricted_ff_enabled(domain):
+        return RESTRICT_DATA_SOURCE_REBUILD.enabled(domain)
```

Comment:
Can we memoize `RESTRICT_DATA_SOURCE_REBUILD.get_enabled_domains()` so we do this only once?

Reply:
Good point